### PR TITLE
Support Blade 15 Base (Early 2020) #356

### LINKED ISSUE #356: Razer-macos 0.4.1
Work under Razer Blade Base 2020 but do nothing - device list is empty.
Version master not work - java issue.

### PR DESCRIPTION
**Pull request by dhobi**
Added support for Blade 15 Base (Early 2020) 
This will fix #356 

Note: This is untested. Could @LuxuryWasTaken have a look at this?
HowTo:

1. Download code on this branch here: https://github.com/1kc/razer-macos/tree/Support_Blade_15_Base_(Early_2020)_%23356
2. yarn
3. yarn dev

### CODE REVIEW

**Comment by 1kc** [comment]
Amazing! @LuxuryWasTaken I've made a test release [here](https://github.com/1kc/razer-macos/releases/tag/v0.4.3.test) for you to download, just to make it easier.

@dhobi don't forget to submit this PR to https://issuehunt.io/r/1kc/razer-macos/issues/356, don't think we've actually tested that the reward thing actually works so please give it a go.

**Comment by dhobi** [comment]
It would be interesting for me to know if the brightness on 100% is still just around 40% of the real brightness (100/2.55). 
My current guess is that we have to update the "blade brightness" methods as well.

**Comment by LuxuryWasTaken** [comment]
Yes, it's working, but like I said something must be changed related with color picker cause it's not working. But you can use / select colors from menu bar (I really don't know a lot about macOS, in Windows, we're calling it as Taskbar)

Also, brightness is still buggy but like I said, Razer Blade is coming with brightness keys in function keys and it works perfectly, so I guess there is no sense of adding it as feature.

**Comment by dhobi** [comment]
> Yes, it's working, but like I said something must be changed related with color picker cause it's not working. 

Could you describe this a bit in more detail? The color picker is just there to "set/save" the custom color for a device. This custom color has to be selected in the menu / tray / taskbar afterwards.

> Also, brightness is still buggy ...

I hope this won't be the case anymore with the latest commit here.

**Comment by LuxuryWasTaken** [comment]
This is my first time using this software, I guess color picker is working. Is it supposed to be working like this?
1)Save a color from color picker
2)Click to Razer macOS in the menu bar
3)Under Razer Blade, hover your mouse to Static
4)Select custom color.

If it's working like that normally, it's working yes. I thought color sets immediately when you select from color picker.

Also, I guess some of effects are not working, I can go to windows and test all of them if you want.

**Comment by dhobi** [comment]
> If it's working like that normally, it's working yes. I thought color sets immediately when you select from color picker.

Yes, this is how it currently works.

> Also, I guess some of effects are not working, I can go to windows and test all of them if you want.

Sure, that would be great. We could either set them as "missing" or try to fix it.

**Comment by LuxuryWasTaken** [comment]
gg

**Comment by dhobi** [comment]
Hmm what does the picture show exactly? 

It would be great to have a list like this:
- available: Audio Meter, Breathing, Spectrum Cycling, Static, Ripple, ...
- working: Static, ...
- not working: Breathing, Spectrum Cycling, Ripple, ...

**Comment by LuxuryWasTaken** [comment]
only available audio meter, breathing, spectrum cycling, static.

I'm switching to mac to confirm which is working.

but I don't think all of them necessarry, eg: audio meter is suck because chroma is not per key for base laptops.

**Comment by LuxuryWasTaken** [comment]
WHAT THE FUCK, TRIED WHEEL AND IT'S WORKING
THE PER KEY THING WAS RAZER SYNAPSE'S LIMITATION?

**Comment by LuxuryWasTaken** [comment]
Also, in menu bar, brightness is showing 255%, so yes your thought is true.

**Comment by LuxuryWasTaken** [comment]
Not working: Wave, Reactive, Starlight, Ripple.
Just wondering, can you add support for wave?

I also have a lot of devices for Razer

Deathadder Elite
Basilisk Ultimate Wireless with Dock
Blackshark v2
Blackwidow Elite
Blackwidow v3 Pro
Deathadder V2 Pro wireless
Lmk if you need testing for those devices.

**Comment by LuxuryWasTaken** [comment]
So I went to OpenRazer's git and asked, they sent this code.
` class RazerBladeEarly2020Base(_RippleKeyboard): 
     """ 
     Class for the Razer Blade Base (Early 2020) 
     """ 
     EVENT_FILE_REGEX = re.compile(r'.*Razer_Blade(-if01)?-event-kbd') 
  
     USB_VID = 0x1532 
     USB_PID = 0x0255 
     HAS_MATRIX = True 
     MATRIX_DIMS = [1, 16] 
     METHODS = ['get_device_type_keyboard', 'set_wave_effect', 'set_static_effect', 'set_spectrum_effect', 
                'set_reactive_effect', 'set_none_effect', 'set_breath_random_effect', 'set_breath_single_effect', 
                'set_breath_dual_effect', 'set_custom_effect', 'set_key_row', 
                'set_starlight_random_effect', 'set_starlight_single_effect', 'set_starlight_dual_effect', 
                'set_ripple_effect', 'set_ripple_effect_random_colour'] 
  
     DEVICE_IMAGE = "https://assets2.razerzone.com/images/blade-15/blade-15-base-model-spec-image-v2.png" 
`
     **MATRIX_DIMS = [1, 16]** 
According to this class, Razer limited it to 1 row, 16 column. So basically, it's not per zone, it's per column customizable RGB.

TL;DR: 
Wave effect can apply according to their port.

**Comment by dhobi** [comment]
I'm aware of the code in the openrazer project. The changes done on this PR are based on exactly this code fragment you posted (+some in .c files).

I can't explain why it works in openrazer but doesn't in Razer macOS...

**Comment by 1kc** [comment]
Unfortunately it's pretty hard to debug without @dhobi actually having a Blade 15 @LuxuryWasTaken.
If we would to dig deeper into the cause, we would have to use usb packet capture to see the differences in packets sent and received between Razer MacOS and openrazer. This most definitely requires having the device.
I think the progress is pretty good given the goal is provide best effort support for razer devices.

**Comment by LuxuryWasTaken** [comment]
hey, I just checked razer macOS animations and didn't see Wave effect, maybe that's why it's not working for blade?

**Comment by dhobi** [comment]
Which program do you mean with razer macOS? This software here? I think we show the wave effect (as the openrazer implementation does as well) but as you say it does not work at all...

**Review by WrathofthePast — CHANGES_REQUESTED** [review]
Your changes to razerchromacommon.c are already present in 1kc/librazermacos. You can delete that file from this PR. Furthermore, the last line you added to razerkbd_driver.c is already present in librazermacos.

If you move the remaining changes on razerkbd_driver.c & .h to a PR on librazermacos and delete it from this PR, it'll will be able to be automatically merged.
